### PR TITLE
refactor(`crosschain`): move ballot voting logic into `observer`

### PR DESCRIPTION
**Pull request by lumtis**
# Description

Move the voting logic for inbound and outbound to the `observer` module and add unit tests to the voting functions.

Closes: https://github.com/zeta-chain/node/issues/1451

### CODE REVIEW

**Comment by brewmaster012** [comment]
similar issue:
https://github.com/zeta-chain/node/issues/1515

File: x/observer/keeper/vote_outbound.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		return false, false, ballot, "", observertypes.ErrInvalidStatus
+	}
+
+	// check if voter is authorized
+	if ok := k.IsAuthorized(ctx, voter); !ok {
+		return false, false, ballot, "", observertypes.ErrNotObserver
+	}
+
+	// fetch or create ballot
+	ballot, isNew, err = k.FindBallot(ctx, ballotIndex, observationChain, observertypes.ObservationType_OutBoundTx)
```
Comment (kingpinXD) [inline x/observer/keeper/vote_outbound.go:39]: Since we are already refactoring some of the ballot logic, it would be good to add the `voter list` and `threshold` as part of the parameter list so that the TSS voter can also use this function.

This can, however, be done as a separate PR, too, if you want to limit the changes on this one.

Comment (lumtis) [reply x/observer/keeper/vote_outbound.go:39]: I might think this is a bit related to the first comment? https://github.com/zeta-chain/node/issues/1806

**Review by ws4charlie — APPROVED** [review]
looks good

File: x/observer/keeper/vote_outbound_test.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+			observer,
+		)
+		require.NoError(t, err)
+
+		// ballot should be finalized since there is only one observer
+		require.False(t, isFinalized)
+		require.False(t, isNew)
+		expectedBallot, found := k.GetBallot(ctx, "index")
+		require.True(t, found)
+		require.Equal(t, expectedBallot, ballot)
```
Comment (kingpinXD) [inline x/observer/keeper/vote_outbound_test.go:191]: maybe we can add a test in which the ballot is not finalized , but vote gets added . 
We can set the observer set to a higher number

File: x/observer/keeper/vote_inbound_test.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+			common.CoinType_ERC20,
+			observer,
+			"index",
+			"inTxHash",
+		)
+		require.NoError(t, err)
+
+		// ballot should not be finalized as the threshold is not reached
+		require.False(t, isFinalized)
+		require.False(t, isNew)
```
Comment (kingpinXD) [inline x/observer/keeper/vote_inbound_test.go:321]: Same comment as the one for outbound TX , 
Maybe we can add unit test which adds vote but doesn't finalize the ballot

**Review by brewmaster012 — APPROVED** [review]
LGTM

File: x/crosschain/keeper/msg_server_vote_inbound_tx.go [hunk 4]
```diff
@@ -57,74 +55,60 @@ import (
...
-	}
-
 	index := msg.Digest()
-	// Add votes and Set Ballot
-	// GetBallot checks against the supported chains list before querying for Ballot
-	ballot, isNew, err := k.zetaObserverKeeper.FindBallot(ctx, index, observationChain, observationType)
+
+	// vote on inbound ballot
+	// use a temporary context to not commit any ballot state change in case of error
+	tmpCtx, commit := ctx.CacheContext()
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/crosschain/keeper/msg_server_vote_inbound_tx.go:62]: Added it to make https://github.com/zeta-chain/node/blob/cfd27a705ab5d299e8eaa0e078bd6f2360a80ab2/x/crosschain/keeper/msg_server_vote_inbound_tx_test.go#L150 passing.
It is not necessary in practice now since the message would cancel the entire state transition but might become if we use the message internally. So it keep the specs: if the inbound finalized check doesn't pass, no ballot state change

File: x/observer/keeper/vote_inbound_test.go (hunk 3 above)
Comment (lumtis) [reply x/observer/keeper/vote_inbound_test.go:321]: You mean the opposite? The vote is added to an existing ballot but it is not finalized?

Comment (lumtis) [reply x/observer/keeper/vote_inbound_test.go:321]: Looks good to me to add it. We'll need to add the extensive testing for voting logic here: https://github.com/zeta-chain/node/issues/1806

Comment (lumtis) [reply x/observer/keeper/vote_inbound_test.go:321]: e0395310

Comment (lumtis) [reply x/observer/keeper/vote_inbound_test.go:321]: Oh, I see, ballot created but not finalized? I can add this as well

Comment (lumtis) [reply x/observer/keeper/vote_inbound_test.go:374]: [580e897](https://github.com/zeta-chain/node/pull/1511/commits/580e897f52e5e630746686c06567784f9c00064a)

File: x/observer/keeper/vote_outbound_test.go (hunk 2 above)
Comment (lumtis) [reply x/observer/keeper/vote_outbound_test.go:237]: [580e897](https://github.com/zeta-chain/node/pull/1511/commits/580e897f52e5e630746686c06567784f9c00064a)

**Review by kingpinXD — APPROVED** [review]
